scripts/cybershake/cybershake_progress.py

```diff
@@ -0,0 +1,316 @@
+    grouped_df.sort_index(axis=0, level=0, inplace=True)
+
+    # Sanity check
+    assert np.all(
+        grouped_df.index.values == fault_names
+    ), "The fault names of the estimation results and those from the list aren't matching"
+
+    # Create progress dataframe
+    column_t = []
+    for proc_type in PROCESS_TYPES:
+        column_t.append((proc_type, EST_CORE_HOURS_COL))
+        column_t.append((proc_type, ACT_CORE_HOURS_COL))
+    column_t.append(("total", EST_CORE_HOURS_COL))
+    column_t.append(("total", ACT_CORE_HOURS_COL))
+
+    progress_df = pd.DataFrame(
+        index=grouped_df.index.values, columns=pd.MultiIndex.from_tuples(column_t)
+    )
+
+    # Get actual core hours for all faults
+    chours_df = get_chours_used(faults)
+
+    # Populate progress dataframe with estimation data and actual data
+    for proc_type in PROCESS_TYPES:
+        progress_df[proc_type, EST_CORE_HOURS_COL] = grouped_df[
+            proc_type, const.MetadataField.core_hours.value
+        ]
+        progress_df[proc_type, ACT_CORE_HOURS_COL] = chours_df[proc_type]
+
+    # Compute total estimated time and actual time across all faults
+    idx = pd.IndexSlice
+    progress_df["total", EST_CORE_HOURS_COL] = progress_df.loc[
+        :, idx[:, EST_CORE_HOURS_COL]
+    ].sum(axis=1)
+    progress_df["total", ACT_CORE_HOURS_COL] = progress_df.loc[
+        :, idx[:, ACT_CORE_HOURS_COL]
+    ].sum(axis=1)
+    progress_df["total", R_COUNT_COL] = r_counts
+
+    return progress_df
+
+
+def load_progress_df(file: str):
+    return pd.read_csv(file, index_col=[0], header=[0, 1])
+
+
+def write_progress(progress_df: pd.DataFrame, outfile: str, mgmtdb: MgmtDB):
+    """Writes the progress dataframe to outfile (in a nice format)."""
+
+    f = open(outfile, "w")
+
+    def get_usage_str(fault: str, proc_type: str):
+        return "{:.2f}/{:.2f}".format(
+            progress_df.loc[fault, (proc_type, ACT_CORE_HOURS_COL)],
+            progress_df.loc[fault, (proc_type, EST_CORE_HOURS_COL)],
+        )
+
+    to_print_mask = np.ones(progress_df.shape[0], dtype=bool)
+
+    f.write(
+        _fault_template_fmt.format(
+            "Fault name",
+            "LF time",
+            "rels",
+            "C",
+            "HF time",
+            "rels",
+            "C",
+            "BB time",
+            "rels",
+            "C",
+            "Total",
+        )
+    )
+    f.write("\n")
+
+    est_completed_all = {}
+    total_complete_rels = {}
+    for fault in progress_df.index.values[to_print_mask]:
+        est_completed_all[fault] = {}
+        num_all_rels = int(
+            progress_df.loc[fault, ("total", R_COUNT_COL)]
+        )  # not sure why this is np.float64
+        num_complete_rels = {}
+        usage_str = {}
+        is_complete = {}
+        for proc_type in PROCESS_TYPES:
+            usage_str[proc_type] = get_usage_str(fault, proc_type)
+            ncl = mgmtdb.num_task_complete(
```

Comment:
Because it's a standalone script it doesn't matter. But ideally, we wouldn't be fetching data in a print/write function.

Reply:
> Also, in-general. What do we think the use case of the script is. For writing progress to a file or to print it to the console? I personally wouldn't use the file much.

I originally stitched existing codes together to send alert to slack, which prefers things in a file. Short summary to Slack, and when something looks funny, user can log into HPC and have a look at the file.

Reply:
Does that mean for every run we'd have to set automatic runs for the script?

I'd assume that for smaller runs (like Vahid and Mike are doing) they could just run the script in a post-hoc fashion to get a summary of estimated vs actual CH rather than a large run where you run it periodically?

Reply:
We used to have a cron job for this
```
0 21,0,3,6,9,12,15,18 * * * python /home/baes/dashboard/cybershake_progress.py /nesi/nobackup/nesi00213/RunFolder/Cybershake/v20p4  /nesi/nobackup/nesi00213/RunFolder/Cybershake/v20p4/list_excluded_20p4.txt /nesi/nobackup/nesi00213/RunFolder/Cybershake/v20p4/cs_progress.txt --alert 2>/dev/null
```
Of course, they can run this script from terminal without `--alert`

I can print to screen as a default behaviour, redirect stdout to a file when we run - it will be a bit messy to pass some of the output around for `--alert`, but if it improves the usability, I can certainly entertain the idea.

Reply:
In an ideal world we'd make one dataframe/blob of data to be printed, and then pass that to different functions for output (stdout, file, slack alert) depending on arguments.

Reply:
OK - let's keep dreaming :) I'm doing some restructuring now.

Reply:
:thought_balloon:

Reply:
Done the re-structuring. By default, it prints to screen, and data collection vs printing have been fully separated for idealists. I like it now.